File: AgentDeplacement.py
import random as r

from abc import ABC, abstractmethod
from Agent import Agent

class AgentDeplacement(ABC, Agent):
    
    def __init__(self, pos):
        super().__init__(pos)
        self.sexe = r.randint(0, 1)
        self.can_reproduce = False
        
    @abstractmethod
    def move(self):
        pass
    
    def moore_neighborhood(pos: tuple[int, int], grid_size: int):

        """_summary_
        Generate with IA
        Returns: Liste des cases accessibles
            _type_: _description_
        """
        x, y = pos
        neighbors = []
        for dx in [-1, 0, 1]:
            for dy in [-1, 0, 1]:
                if dx == 0 and dy == 0:  # Ignorer la position actuelle
                    continue
                nx, ny = x + dx, y + dy
                if 0 <= nx < grid_size and 0 <= ny < grid_size:
                    neighbors.append((nx, ny))
        return neighbors

File: AgentDeplacement.java
import java.util.ArrayList;
import java.util.Arrays;

public abstract class AgentDeplacement extends Agent{

    private int sexe;
    private Position[] neighbors;
    public static final ArrayList<Integer> DIRECTION = new ArrayList<>(Arrays.asList(-1, 0, 1)); 

    public AgentDeplacement(Position pos) {
        
        super(pos);
        this.sexe = (int)(Math.random() * 2);
        this.neighbors = new Position[8];
    }

    private void mooreNeighborhood(Position pos, int gridSize){

        int indice = 0;

        for (int dx : DIRECTION){
            for (int dy : DIRECTION){
                if (dx == 0 && dy == 0){
                    continue;
                }
                // TODO #1 méthode accessPosition dans Position
                int nx = this.getPos().getX() + dx;
                int ny = this.getPos().getY() + dy;
                if ((nx < gridSize) && (ny < gridSize)){
                    try{
                        neighbors[indice] = new Position(nx, ny, 0);
                        indice++;
                    }catch(IndexOutOfBoundsException e){
                        System.out.println(String.format("%s, indice = %d", e, indice));
                    }
                }
            }
        }
    }
    
    @Override
    public void reproduce() {
        // TODO Auto-generated method stub
        throw new UnsupportedOperationException("Unimplemented method 'reproduce'");
    }

    @Override
    public void reproduceCooldaown() {
        // TODO Auto-generated method stub
        throw new UnsupportedOperationException("Unimplemented method 'reproduceCooldaown'");
    }

    public int getSexe() {
        return sexe;
    }

    public Position[] getNeighbors() {
        return neighbors;
    }

    public abstract void move();
}